man/getxhead.3
.TH GETXHEAD 3 "12 September 1995"
.FB
.SH NAME
getxhead, putxhead \- read and write index header
.SH SYNOPSIS
#include <firstbase/fb.h>
.sp 1
fb_getxhead(fd, bsmax, bsend)
.br
int fd;
.br
long *bsmax, *bsend;
.sp 1
fb_putxhead(fd, bsmax, bsend)
.br
int fd;
.br
long bsmax, bsend;
.SH DESCRIPTION
As explained in the
\fIindex\fP(4) manual page, a database structure contains
all of the elements needed for an entire database, including
file names, descriptors and other variables that define an index.
\fIFb_getxhead\fP and \fIfb_putxhead\fP, along with a few other UNIX
calls, enable the building and maintaining of complex (multi-field)
flat \fIFirstBase\fP indexes.
.PP
\fBNote\fP: the following discussion concerns normal (flat) \fIFirstBase\fP
indexes and does \fBnot\fP apply to btree indexes. See \fIbtree\fP(4) for more
details on the \fIFirstBase\fP Btree+ indexes.
.PP
Formally, \fIfb_getxhead\fP and \fIfb_putxhead\fP read and write the
header numbers needed to maintain a \fIFirstBase\fP index. Note that
\fIfb_getxhead\fP uses two long pointers, i.e. the address to two long
variables. In this manner, \fIfb_getxhead\fP reads two distinct variables
from the index header \fIfd\fP. \fIFb_putxhead\fP writes the two long values
passed in to the index header indicated by \fIfd\fP.
.PP
But, \fIfb_getxhead\fP and \fIfb_putxhead\fP are just two of the pieces
needed to update a complex index. The following discussion covers
the complete index update task.
.PP
When a database is opened \fIwith an index\fP using \fIopendb\fP(3),
or when an index is tied to a database using \fIopenidx\fP(3),
the following variables from the database structure are initialized:
.TP 9
.B ifd
index file descriptor
.TP 9
.B ihfd
index header field descriptor
.TP 9
.B irecsiz
size of an index record
.TP 9
.B bsmax
binary search maximum, the index record count
.TP 9
.B bsend
binary search end point, i.e. the last sorted entry
.TP 9
.B irec
character buffer big enough for one index entry
.TP 9
.B ip
array of field pointers, the sort-by fields
.TP 9
.B afld
temporary space for a single field
.TP 9
.B bfld
temporary space for a single field
.PD
.PP
Each \fIFirstBase\fP index is a fixed length miniature database.
Entering a new index entry into an open index is a two step procedure.
The first step creates the string to be written to the index, while the second
step does the actual write.
.PP
Again, the first step to adding a new index entry
is to create the string to be written to the index.
This index entry, the string that will be written to the index, needs
to be well formed, i.e. it needs to be the exact number of characters or bytes
to fit into an single index entry slot.
The database variable \fIirecsiz\fP contains the total
length of an index entry.
.PP
When a database index is opened, \fIirecsiz\fP is calculated by
adding together the
lengths of the fields being indexed (\fIcdb_ip\fP), plus a constant.
Most of the constant represents the length of the
record pointer, a string that is always FB_RECORDPTR (10) characters long.
An additional one (1) is added to the constant to account for the NEWLINE at
the end of the string. So, the constant added to the sum of the lengths
is eleven (11).
.PP
It is also worth noting that the \fIFirstBase\fP data type of the fields being
written to the index (the sort-by fields) determine whether the field is
left or right justified. Numeric type fields (FB_OFNUMERIC) need to be
right justified. Date fields require the shifting of the year to the front
of the string. The \fIFirstBase\fP library routine \fImakess\fP(3) will
produce the correct search string.
.PP
The following routine will create a proper string to be
written to the index. Note that this routine depends on the
index having been formally tied to the database via \fIopendb\fP(3) or
\fIopenidx\fP(3). If this is not the case, some of the variables from
the list above will need to be provided.
.PP
.nj
.nf
.ft CW
   /*
    * make_entry - make an entry based on the current index
    */

   static make_entry(dp)
      fb_database *dp;

      {
         char buf[MAXLINE];
         int siz, i, ixfields;
         fb_field *fp;

         db->irec[0] = NULL;
         ixfields = dp->ifields - 1;
         for (i = 0; i < MAXBY && i < ixfields; i++){
            fp = dp->ip[i];
            strcpy(dp->afld, fp->fld);
            fb_makess(dp->afld, fp->type, fp->size);

            /* add this value to the index record buffer */
            strcat(dp->irec, dp->afld);
            }

         /* create the record pointer portion of the index record */
         sprintf(dp->afld, "%010ld\n", dp->rec);
         strcat(dp->irec, dp->afld);
      }
.ft
.fi
.ju
.PP
The second step to adding an index entry to locate the proper place for this
index entry to be written. To do this, \fIfb_getxhead\fP(3) is used to read
header information about the index, seek to the the proper byte location,
and write the index record entry. Once written, the index header
is updated to reflect the new index entry.
.PP
The following code fragment does the job. Again, note that this code depends
on the index being formally tied to the database, which may not be the case.
.PP
.nj
.nf
.ft CW
      /* get bsmax, bsend from index header file */
      fb_getxhead(dp->ihfd, &dp->bsmax, &dp->bsend);

      /* seek and write index entry */
      lseek(dp->ifd, dp->bsmax * dp->irecsiz, 0);
      write(dp->ifd, dp->irec, dp->irecsiz);

      /* redo header -- note: only bsmax changed */
      fb_putxhead(dp->ihfd, ++dp->bsmax, dp->bsend);
.ft
.fi
.ju
.PP
Using the above methods, a \fIFirstBase\fP multi-field flat indexes
can be maintained at run time. Remember to close the file descriptors at
some point \(em \fIclosedb\fP(3) will do this for those descriptors formally
tied to the database structure. A sample program named \fIkey_add.c\fP that
puts this all together is in $FIRSTBASEHOME/applications/sample_src.
.PP
The \fIFirstBase\fP tool \fIdbregen\fP(1) can be used to resort an index
and update the index header file without recreating the index (\fIdbigen\fP(1)).
.SH SEE ALSO
btree(4), index(4), opendb(3), openidx(3),
makess(3), endate(3), pad(3), rjustify(3).
.SH DIAGNOSTICS
\fIFb_getxhead\fP and \fIfb_putxhead\fP will return the error code FB_ERROR
if any of the reads, writes, or seeks fail.
An FB_AOK signal is returned on correct completion.
.PP
Note that neither of these routines implements any process locking around the
critical sections, the read and write calls.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/librec/getxhead.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: getxhead.c,v 9.0 2001/01/09 02:56:58 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Getxhead_sid[] = "@(#) $Id: getxhead.c,v 9.0 2001/01/09 02:56:58 john Exp $";
#endif

#include <fb.h>
#include <fb_ext.h>

extern short int cdb_dbase_byteorder;
extern short int cdb_cpu_byteorder;

/*
 * getxhead - get two FB_SLONG values from the start of the header.
 */
 
   fb_getxhead(fd, v1, v2)
      int fd;
      long *v1, *v2;

      {
         char buf[FB_SLONG+FB_SLONG+3];
         int rlen = FB_SLONG + FB_SLONG + 1;	/* 1 to get past newline */;

	 if (lseek(fd, (long) FB_HEADSTART, 0) < 0L){
	    fb_lerror(FB_SEEK_ERROR, SYSMSG[S_BAD_INDEX], NIL);
	    return(FB_ERROR);
	    }
	 if (read(fd, buf, (unsigned) rlen) != rlen){
	    fb_lerror(FB_READ_ERROR, SYSMSG[S_BAD_INDEX], NIL);
	    return(FB_ERROR);
	    }
         memcpy((char *) v1, buf, FB_SLONG);
         memcpy((char *) v2, buf + FB_SLONG, FB_SLONG);
         if (cdb_dbase_byteorder != cdb_cpu_byteorder){
            M_32_SWAP((*v1));
            M_32_SWAP((*v2));
            }
         /*
          * did do a seek here to get past newline.
          * now, newline is read at same time into buf above.
          * lseek(fd, 1L, 1);
          */
	 return(FB_AOK);
      }

/*
 * skipxhead - skip past the index header
 */
 
   fb_skipxhead(fd)
      int fd;

      {
         long rlen = FB_SLONG + FB_SLONG + 1 + FB_HEADSTART;
         					/* 1 to get past newline */;

	 if (lseek(fd, rlen, 0) < 0L){
	    fb_lerror(FB_SEEK_ERROR, SYSMSG[S_BAD_INDEX], NIL);
	    return(FB_ERROR);
	    }
	 return(FB_AOK);
      }
